amundsen_application/static/js/ducks/feedback/tests/index.spec.ts

```diff
@@ -0,0 +1,131 @@
+  ResetFeedback,
+} from '../types';
+
+
+describe('feedback ducks', () => {
+  let formData: FormData;
+  beforeAll(() => {
+    formData = new FormData();
+    const testData = { rating: 10, comment: 'This is a test' };
+    Object.keys(testData).forEach(key => formData.append(key, testData[key]));
+  });
+
+  describe('actions', () => {
+    it('submitFeedback - returns the action to submit feedback', () => {
+      expect(submitFeedback(formData)).toEqual({
+        type: SubmitFeedback.REQUEST,
+        payload: {
+          data: formData
+        },
+      });
+    });
+
+    it('submitFeedbackFailure - returns the action to process failure', () => {
+      expect(submitFeedbackFailure()).toEqual({
+        type: SubmitFeedback.FAILURE,
+      });
+    });
+
+    it('submitFeedbackSuccess - returns the action to process success', () => {
+      expect(submitFeedbackSuccess()).toEqual({
+        type: SubmitFeedback.SUCCESS,
+      });
+    });
+
+    it('resetFeedback - returns the action to reset feedback', () => {
+      expect(resetFeedback()).toEqual({
+        type: ResetFeedback.REQUEST,
+      });
+    });
+  });
+
+  describe('reducer', () => {
+    let testState: FeedbackReducerState;
+    beforeAll(() => {
+      testState = { sendState: SendingState.IDLE };
+    });
+
+    it('should return the existing state if action is not handled', () => {
+      expect(reducer(testState, { type: 'INVALID.ACTION' })).toEqual(testState);
+    });
+
+    it('should handle SubmitFeedback.REQUEST', () => {
+      expect(reducer(testState, submitFeedback(formData))).toEqual({ sendState: SendingState.WAITING });
+    });
+
+    it('should handle SubmitFeedback.SUCCESS', () => {
+      expect(reducer(testState, submitFeedbackSuccess())).toEqual({ sendState: SendingState.COMPLETE });
+    });
+
+    it('should handle SubmitFeedback.FAILURE', () => {
+      expect(reducer(testState, submitFeedbackFailure())).toEqual({ sendState: SendingState.ERROR });
+    });
+
+    it('should handle ResetFeedback.REQUEST', () => {
+      expect(reducer(testState, resetFeedback())).toEqual({ sendState: SendingState.IDLE });
+    });
+  });
+
+  describe('sagas', () => {
+    let action: SubmitFeedbackRequest;
+    beforeAll(() => {
+      action = submitFeedback(formData);
+    });
+
+    describe('submitFeedbackWatcher', () => {
+      it('takes every SubmitFeedback.REQUEST with submitFeedbackWorker', () => {
+        testSaga(submitFeedbackWatcher)
+          .next()
+          .takeEvery(SubmitFeedback.REQUEST, submitFeedbackWorker);
+      });
+    });
+
+    describe('submitFeedbackWorker', () => {
+      it('executes submit feedback flow', () => {
+        testSaga(submitFeedbackWorker, action)
+          .next()
+          .call(feedbackSubmit, formData)
+          .next()
+          .put(submitFeedbackSuccess())
+          .next()
+          .delay(2000)
+          .next()
+          .put(resetFeedback())
+          .next()
+          .isDone();
```

Comment:
This type of test seems very cryptic and tightly coupled with the code structure. I'm not sure how much value this adds overall.

Reply:
Sagas are iterators, so the way I understand that they should be tested is to verify that each iteration step proceeds as expected. This is also how redux-saga has documented to test sagas:
https://redux-saga.js.org/docs/advanced/Testing.html
> There are two main ways to test Sagas: testing the saga generator function step-by-step or running the full saga and asserting the side effects.

I've chosen the step by step testing as this seemed to be the approach that required the least amount of burden in terms of learning curve.

Reply:
It's also worth noting our sagas are currently extremely straight forward, they do a thing then return a thing. Our logic is purely linear and doesn't have any complicated branching logic yet or leverage some of the more complicated saga features.